### PR TITLE
AI fills available docks like human players do

### PR DESCRIPTION
**Pull request by FS-21**
In YR when AI has multiple Factories it clones the built object.

Ares fix the behavior with a new tag:
[GlobalControls]
AllowParallelAIQueues=no

But in the case of aircraft factories they have a bug that only fills all the docks of the primary factory, limiting the AI gameplay.

Now AI can fill all available docks like human players do.

### CODE REVIEW

**Comment by NetsuNegi** [comment]
Nice job!!!!!!!!!!!!!!

File: src/Ext/Building/Body.cpp [hunk 1]
```diff
@@ -3,6 +3,142 @@
...
+	auto pOwner = pThis->Owner;
+
+	if (!pOwner || pOwner->ProducingAircraftTypeIndex < 0)
+		return;
+
+	auto BuildingExt = BuildingExt::ExtMap.Find(pThis);
+	if (!BuildingExt
+		|| pOwner->Type->MultiplayPassive
+		|| pOwner->IsNeutral())
+		return;
```
Comment (Metadorius) [inline src/Ext/Building/Body.cpp:17]: Lacks braces. Check styleguide.

File: src/Ext/Building/Body.cpp [hunk 2]
```diff
@@ -3,6 +3,142 @@
...
+
+	// Obtain a list of air factories for optimizing the comparisons
+	for (auto pBuilding : pOwner->Buildings)
+	{
+		if (pBuilding->Type->Factory == AbstractType::AircraftType)
+		{
+			if (!currFactory && pBuilding->Factory)
+			{
+				currFactory = pBuilding->Factory;
+			}
```
Comment (Metadorius) [inline src/Ext/Building/Body.cpp:44]: Redundant braces

File: src/Ext/Building/Body.cpp [hunk 3]
```diff
@@ -3,6 +3,142 @@
...
+		int nOccupiedDocks = BuildingExt::CountOccupiedDocks(pBuilding);
+
+		if (nOccupiedDocks < nDocks)
+		{
+			return true;
+		}
+		else
+		{
+			return false;
+		}
```
Comment (Metadorius) [inline src/Ext/Building/Body.cpp:136]: Braces again

File: src/Ext/Building/Hooks.cpp [hunk 4]
```diff
@@ -32,4 +32,40 @@ DEFINE_HOOK(0x449ADA, BuildingClass_MissionConstruction_DeployToFireFix, 0x0)
...
+
+	HouseClass* pOwner = pBuilding->Owner;
+
+	if (!pOwner)
+		return 0;
+
+	if (pOwner->Type->MultiplayPassive
+		|| pOwner->IsPlayer()
+		|| pOwner->IsNeutral())
+		return 0;
```
Comment (Metadorius) [inline src/Ext/Building/Hooks.cpp:55]: Code duplication here?

File: src/Phobos.cpp [hunk 5]
```diff
@@ -217,6 +218,7 @@ DEFINE_HOOK(0x66E9DF, RulesClass_Process_Phobos, 0x8)
 	GET(CCINIClass*, rulesINI, EDI);
 
 	Phobos::Config::DevelopmentCommands = rulesINI->ReadBool("GlobalControls", "DebugKeysEnabled", Phobos::Config::DevelopmentCommands);
+	Phobos::Config::AllowParallelAIQueues = rulesINI->ReadBool("GlobalControls", "AllowParallelAIQueues", Phobos::Config::AllowParallelAIQueues);
```
Comment (Metadorius) [inline src/Phobos.cpp:221]: Add comment that states that this is an Ares tag.

**Review by Metadorius — CHANGES_REQUESTED** [review]

File: src/Ext/Building/Hooks.cpp (hunk 4 above)
Comment (FS-21) [reply src/Ext/Building/Hooks.cpp:55]: I removed pOwner->Type->MultiplayPassive & pOwner->IsNeutral() in the main function for reducing duplicated checks.

**Review by FS-21 — COMMENTED** [review]
Updated code style due to the review

File: src/Ext/Building/Body.cpp [hunk 6]
```diff
@@ -3,6 +3,139 @@
...
+	auto pOwner = pThis->Owner;
+
+	if (!pOwner || pOwner->ProducingAircraftTypeIndex < 0)
+		return;
+
+	auto BuildingExt = BuildingExt::ExtMap.Find(pThis);
+	if (!BuildingExt)
+	{
+		return;
+	}
```
Comment (Metadorius) [inline src/Ext/Building/Body.cpp:17]:
```suggestion
		return;
```

File: src/Ext/Building/Body.cpp [hunk 7]
```diff
@@ -3,6 +3,139 @@
...
+		return 0;
+
+	int nOccupiedDocks = 0;
+
+	for (auto i = 0; i < pBuilding->RadioLinks.Capacity; ++i)
+	{
+		if (auto const pLink = pBuilding->GetNthLink(i))
+		{
+			nOccupiedDocks++;
+		}
```
Comment (Metadorius) [inline src/Ext/Building/Body.cpp:114]:
```suggestion
			nOccupiedDocks++;
```

**Review by Metadorius — APPROVED** [review]

**Review by Metadorius — APPROVED** [review]

**Pull request merged**